### ISSUE TITLE
avoid depending on child process execution when the target OS does not support it

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#10750]
There are a few places in the Zig compiler where we invoke a child process, but none of them are *strictly* necessary. This issue is to detect when the OS does not support child process execution and produce a build of the compiler that does not attempt to use child processes.

For now the detection logic can be simply:

```zig
const can_spawn = switch (builtin.os.tag) {
    .wasi => false,
    else => true,
};
```

Here is a complete list of everywhere Zig might want to spawn a child process:

 * `zig test` - instead print a message about not being able to spawn child processes and exit success.
 * `zig run` - instead print "unable to spawn child processes" and exit failure.
 * for clang we can invoke clang_main directly instead of as a child process
 * for lld we can invoke the lld library functions instead of as child process
 * same thing for `zig ar`, `zig dlltool`, etc.
 * for detecting native libc, just return error.LibcNotDetected or whatever the appropriate error is.

This is a prerequisite for #10716.

**Comment by marler8997** [comment]
I'd be curious to see how significant the performance cost is for sub processes on platforms that support it.  With this in place we could easily find out!

**Comment by zigazeljko** [comment]
Doesn't `zig build` also spawn child processes?

**Comment by zseri** [comment]
Why should `zig test` exit with success?

**Comment by marler8997** [comment]
> Doesn't zig build also spawn child processes?

Oh yes. To compile anything zig build will spawn a subprocess like `zig build-exe ...`.

**Comment by matu3ba** [comment]
This should probably be first implemented before #1356 with potentially panicking subprocess being started from the subprocess that supervises test results.

**Comment by andrewrk** [comment]
Landed in 21135387fb7c2dbaf70a72f2c97341e9c1307045
